File: webservice-product-invisible-on-storefront/python/repair_invisible_product.py
"""Detect and repair PrestaShop products created via webservice that are invisible
on the storefront despite showing active in the back office.

The full admin product save wires up category_product links, shop associations,
and search index rows as side effects of the whole controller save chain. The
webservice Product::add()/update() path only writes what the submitted resource
body explicitly includes. A payload that sets active=1 without an
associations.categories block carrying id_category_default, or without an
associations.shops entry, leaves the product active in product/product_shop but
with no category link and no shop association, so front-end catalog queries that
join through those tables never return it (PrestaShop/PrestaShop issues #15317
and #28409).

This script lists recently created active products with display=full, inspects
the associations block already returned, cross-checks id_category_default against
real categories, and flags or repairs the missing links. Repair merges the fix
onto the full current resource and PUTs it back, then re-GETs to confirm. A
product whose default category itself is invalid is only ever flagged, never
auto-written, since guessing a replacement category could mis-file it.

Run on a schedule. Safe to run again and again.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("repair_invisible_product")

PRESTASHOP_URL = os.environ["PRESTASHOP_URL"].rstrip("/")
PRESTASHOP_WS_KEY = os.environ["PRESTASHOP_WS_KEY"]
EXPECTED_SHOP_IDS = [int(s) for s in os.environ.get("EXPECTED_SHOP_IDS", "1").split(",") if s.strip()]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
AUTH = (PRESTASHOP_WS_KEY, "")


def decide_product_repair(product, context):
    """Pure decision function, no I/O.

    product: {active: 0|1, visibility: "both"|"catalog"|"search"|"none",
              id_category_default: int, associations: {categories: [int], shops: [int]}}
    context: {expectedShopIds: [int], validCategoryIds: [int]}

    Returns {status: "ok"|"needs_repair"|"unrepairable", missing: [str], patch: dict|None}.
    """
    if product["active"] != 1:
        return {"status": "ok", "missing": [], "patch": None}

    missing = []
    categories = product["associations"]["categories"]
    id_category_default = product["id_category_default"]

    if len(categories) == 0:
        missing.append("categories")
    elif id_category_default not in categories:
        missing.append("id_category_default_not_in_categories")

    shops = product["associations"]["shops"]
    expected_shop_ids = context["expectedShopIds"]
    if len(shops) == 0 or not any(sid in shops for sid in expected_shop_ids):
        missing.append("shops")

    if product["visibility"] == "none":
        missing.append("visibility")

    if id_category_default not in context["validCategoryIds"]:
        missing.append("default_category_invalid")
        return {"status": "unrepairable", "missing": missing, "patch": None}

    if not missing:
        return {"status": "ok", "missing": [], "patch": None}

    patch = {}
    if "categories" in missing or "id_category_default_not_in_categories" in missing:
        patch["associations"] = {
            "categories": sorted(set(categories) | {id_category_default})
        }
    if "shops" in missing:
        patch.setdefault("associations", {})["shops"] = list(expected_shop_ids)
    if "visibility" in missing:
        patch["visibility"] = "both"

    return {"status": "needs_repair", "missing": missing, "patch": patch}


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(f"{PRESTASHOP_URL}/api/{path}", params=params, auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def api_put(path, resource_key, body):
    r = requests.put(
        f"{PRESTASHOP_URL}/api/{path}",
        params={"output_format": "JSON"},
        auth=AUTH,
        json={resource_key: body},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def list_recent_active_products(date_from, date_to, limit=100):
    data = api_get("products", params={
        "display": "full",
        "filter[active]": 1,
        "filter[date_add]": f"[{date_from},{date_to}]",
        "limit": limit,
    })
    return data.get("products") or []


def category_is_valid(id_category):
    try:
        data = api_get(f"categories/{id_category}")
    except requests.HTTPError:
        return False
    category = data.get("category") or {}
    return str(category.get("active", "0")) == "1"


def get_full_product(id_product):
    data = api_get(f"products/{id_product}", params={"display": "full"})
    return data["product"]


def merge_patch_onto_resource(full_product, patch):
    merged = dict(full_product)
    if "associations" in patch:
        merged["associations"] = {**merged.get("associations", {}), **patch["associations"]}
    if "visibility" in patch:
        merged["visibility"] = patch["visibility"]
    return merged


def put_product(id_product, merged_product):
    return api_put(f"products/{id_product}", "product", merged_product)


def to_decision_shape(product):
    associations = product.get("associations") or {}
    categories = [c["id"] for c in (associations.get("categories") or {}).get("category", [])]
    shops = [s["id"] for s in (associations.get("shops") or {}).get("shop", [])]
    return {
        "active": int(product.get("active", 0)),
        "visibility": product.get("visibility", "both"),
        "id_category_default": int(product.get("id_category_default", 0)),
        "associations": {"categories": categories, "shops": shops},
    }


def run(date_from="2026-07-01", date_to="2026-07-11"):
    flagged = 0
    repaired = 0
    unrepairable = 0
    valid_category_cache = {}

    for raw_product in list_recent_active_products(date_from, date_to):
        id_product = raw_product["id"]
        product = to_decision_shape(raw_product)
        id_category_default = product["id_category_default"]

        if id_category_default not in valid_category_cache:
            valid_category_cache[id_category_default] = category_is_valid(id_category_default)
        valid_category_ids = [cid for cid, ok in valid_category_cache.items() if ok]

        decision = decide_product_repair(product, {
            "expectedShopIds": EXPECTED_SHOP_IDS,
            "validCategoryIds": valid_category_ids,
        })

        if decision["status"] == "ok":
            continue

        flagged += 1
        log.warning("Product %s status=%s missing=%s", id_product, decision["status"], decision["missing"])

        if decision["status"] == "unrepairable":
            unrepairable += 1
            log.error("Product %s has an invalid id_category_default=%s, needs a human to pick a category.",
                       id_product, id_category_default)
            continue

        if DRY_RUN:
            log.info("Dry run. Would PUT products/%s with patch=%s", id_product, decision["patch"])
            continue

        full_product = get_full_product(id_product)
        merged = merge_patch_onto_resource(full_product, decision["patch"])
        put_product(id_product, merged)

        confirm_raw = get_full_product(id_product)
        confirm = to_decision_shape(confirm_raw)
        confirm_decision = decide_product_repair(confirm, {
            "expectedShopIds": EXPECTED_SHOP_IDS,
            "validCategoryIds": valid_category_ids,
        })
        if confirm_decision["status"] == "ok":
            repaired += 1
            log.info("Repaired product %s.", id_product)
        else:
            log.error("Product %s still needs_repair after PUT, missing=%s. Not retrying silently.",
                       id_product, confirm_decision["missing"])

    log.info("Done. %d flagged, %d repaired, %d unrepairable.", flagged, repaired, unrepairable)


if __name__ == "__main__":
    run()

File: webservice-product-invisible-on-storefront/node/repair-invisible-product.js
/**
 * Detect and repair PrestaShop products created via webservice that are invisible
 * on the storefront despite showing active in the back office.
 *
 * The full admin product save wires up category_product links, shop associations,
 * and search index rows as side effects of the whole controller save chain. The
 * webservice Product::add()/update() path only writes what the submitted resource
 * body explicitly includes. A payload that sets active=1 without an
 * associations.categories block carrying id_category_default, or without an
 * associations.shops entry, leaves the product active in product/product_shop but
 * with no category link and no shop association, so front-end catalog queries that
 * join through those tables never return it (PrestaShop/PrestaShop issues #15317
 * and #28409).
 *
 * This script lists recently created active products with display=full, inspects
 * the associations block already returned, cross-checks id_category_default against
 * real categories, and flags or repairs the missing links. Repair merges the fix
 * onto the full current resource and PUTs it back, then re-GETs to confirm. A
 * product whose default category itself is invalid is only ever flagged, never
 * auto-written, since guessing a replacement category could mis-file it.
 *
 * Guide: https://www.allanninal.dev/prestashop/webservice-product-invisible-on-storefront/
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/+$/, "");
const PRESTASHOP_WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const EXPECTED_SHOP_IDS = (process.env.EXPECTED_SHOP_IDS || "1")
  .split(",").map((s) => s.trim()).filter(Boolean).map(Number);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

function basicAuthHeader() {
  return "Basic " + Buffer.from(`${PRESTASHOP_WS_KEY}:`).toString("base64");
}

/**
 * Pure decision function, no I/O.
 *
 * product: {active: 0|1, visibility: "both"|"catalog"|"search"|"none",
 *           id_category_default: number, associations: {categories: number[], shops: number[]}}
 * context: {expectedShopIds: number[], validCategoryIds: number[]}
 *
 * Returns {status: "ok"|"needs_repair"|"unrepairable", missing: string[], patch: object|null}.
 */
export function decideProductRepair(product, context) {
  if (product.active !== 1) {
    return { status: "ok", missing: [], patch: null };
  }

  const missing = [];
  const { categories, shops } = product.associations;
  const idCategoryDefault = product.id_category_default;

  if (categories.length === 0) {
    missing.push("categories");
  } else if (!categories.includes(idCategoryDefault)) {
    missing.push("id_category_default_not_in_categories");
  }

  const expectedShopIds = context.expectedShopIds;
  if (shops.length === 0 || !expectedShopIds.some((id) => shops.includes(id))) {
    missing.push("shops");
  }

  if (product.visibility === "none") {
    missing.push("visibility");
  }

  if (!context.validCategoryIds.includes(idCategoryDefault)) {
    missing.push("default_category_invalid");
    return { status: "unrepairable", missing, patch: null };
  }

  if (missing.length === 0) {
    return { status: "ok", missing: [], patch: null };
  }

  const patch = {};
  if (missing.includes("categories") || missing.includes("id_category_default_not_in_categories")) {
    patch.associations = { categories: [...new Set([...categories, idCategoryDefault])] };
  }
  if (missing.includes("shops")) {
    patch.associations = { ...(patch.associations || {}), shops: expectedShopIds };
  }
  if (missing.includes("visibility")) {
    patch.visibility = "both";
  }

  return { status: "needs_repair", missing, patch };
}

async function apiGet(path, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: basicAuthHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on GET ${path}`);
  return res.json();
}

async function apiPut(path, resourceKey, body) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, {
    method: "PUT",
    headers: { Authorization: basicAuthHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ [resourceKey]: body }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on PUT ${path}`);
  return res.json();
}

async function listRecentActiveProducts(dateFrom, dateTo, limit = 100) {
  const data = await apiGet("products", {
    display: "full",
    "filter[active]": 1,
    "filter[date_add]": `[${dateFrom},${dateTo}]`,
    limit,
  });
  return data.products || [];
}

async function categoryIsValid(idCategory) {
  try {
    const data = await apiGet(`categories/${idCategory}`);
    const category = data.category || {};
    return String(category.active) === "1";
  } catch (err) {
    return false;
  }
}

async function getFullProduct(idProduct) {
  const data = await apiGet(`products/${idProduct}`, { display: "full" });
  return data.product;
}

function mergePatchOntoResource(fullProduct, patch) {
  const merged = { ...fullProduct };
  if (patch.associations) {
    merged.associations = { ...merged.associations, ...patch.associations };
  }
  if (patch.visibility) {
    merged.visibility = patch.visibility;
  }
  return merged;
}

async function putProduct(idProduct, mergedProduct) {
  return apiPut(`products/${idProduct}`, "product", mergedProduct);
}

function toDecisionShape(product) {
  const associations = product.associations || {};
  const categories = ((associations.categories || {}).category || []).map((c) => c.id);
  const shops = ((associations.shops || {}).shop || []).map((s) => s.id);
  return {
    active: Number(product.active || 0),
    visibility: product.visibility || "both",
    id_category_default: Number(product.id_category_default || 0),
    associations: { categories, shops },
  };
}

export async function run(dateFrom = "2026-07-01", dateTo = "2026-07-11") {
  let flagged = 0;
  let repaired = 0;
  let unrepairable = 0;
  const validCategoryCache = new Map();

  for (const rawProduct of await listRecentActiveProducts(dateFrom, dateTo)) {
    const idProduct = rawProduct.id;
    const product = toDecisionShape(rawProduct);
    const idCategoryDefault = product.id_category_default;

    if (!validCategoryCache.has(idCategoryDefault)) {
      validCategoryCache.set(idCategoryDefault, await categoryIsValid(idCategoryDefault));
    }
    const validCategoryIds = [...validCategoryCache.entries()].filter(([, ok]) => ok).map(([id]) => id);

    const decision = decideProductRepair(product, { expectedShopIds: EXPECTED_SHOP_IDS, validCategoryIds });

    if (decision.status === "ok") continue;

    flagged++;
    console.warn(`Product ${idProduct} status=${decision.status} missing=${decision.missing}`);

    if (decision.status === "unrepairable") {
      unrepairable++;
      console.error(`Product ${idProduct} has an invalid id_category_default=${idCategoryDefault}, needs a human to pick a category.`);
      continue;
    }

    if (DRY_RUN) {
      console.log(`Dry run. Would PUT products/${idProduct} with patch=${JSON.stringify(decision.patch)}`);
      continue;
    }

    const fullProduct = await getFullProduct(idProduct);
    const merged = mergePatchOntoResource(fullProduct, decision.patch);
    await putProduct(idProduct, merged);

    const confirmRaw = await getFullProduct(idProduct);
    const confirm = toDecisionShape(confirmRaw);
    const confirmDecision = decideProductRepair(confirm, { expectedShopIds: EXPECTED_SHOP_IDS, validCategoryIds });
    if (confirmDecision.status === "ok") {
      repaired++;
      console.log(`Repaired product ${idProduct}.`);
    } else {
      console.error(`Product ${idProduct} still needs_repair after PUT, missing=${confirmDecision.missing}. Not retrying silently.`);
    }
  }

  console.log(`Done. ${flagged} flagged, ${repaired} repaired, ${unrepairable} unrepairable.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
